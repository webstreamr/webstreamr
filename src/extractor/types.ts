import { Context, UrlResult } from '../types';

export interface Extractor {
  readonly id: string;

  readonly label: string;

  readonly ttl: number;

  readonly supports: (url: URL) => boolean;

  readonly extract: (ctx: Context, url: URL, countryCode: string) => Promise<UrlResult>;
}
