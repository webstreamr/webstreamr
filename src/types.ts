import { Manifest, ManifestConfig } from 'stremio-addon-sdk';

export const TIMEOUT = 'timeout';

export interface Context {
  id: string;
  ip: string;
  config: Config;
  referer?: URL;
}

export type ManifestWithConfig = Manifest & { config: ManifestConfig[] };

export type Config = Partial<Record<CountryCode | 'excludeExternalUrls', string>>;

export type CountryCode = 'de' | 'en' | 'es' | 'fr' | 'it' | 'mx';

export type BlockedReason = 'cloudflare_challenge' | 'unknown';

export interface Meta {
  bytes?: number;
  countryCode: CountryCode;
  height?: number;
  title?: string;
}

export interface UrlResult {
  url: URL;
  isExternal?: boolean;
  error?: unknown;
  label: string;
  sourceId: string;
  ttl?: number;
  meta: Meta;
  requestHeaders?: Record<string, string>;
}
