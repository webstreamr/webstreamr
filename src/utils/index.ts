export * from './embed';
export * from './fetch';
export * from './imdb';
export * from './language';
export * from './log';
export * from './manifest';
export * from './promise';
