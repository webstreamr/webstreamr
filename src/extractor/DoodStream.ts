import randomstring from 'randomstring';
import * as cheerio from 'cheerio';
import { Extractor } from './types';
import { Fetcher } from '../utils';
import { Context, Meta } from '../types';
import { NotFoundError } from '../error';

export class DoodStream implements Extractor {
  readonly id = 'doodstream';

  readonly label = 'DoodStream';

  readonly ttl = 900000; // 15m

  private readonly fetcher: Fetcher;

  constructor(fetcher: Fetcher) {
    this.fetcher = fetcher;
  }

  readonly supports = (_ctx: Context, url: URL): boolean => null !== url.host.match(/dood|do[0-9]go/);

  readonly extract = async (ctx: Context, url: URL, meta: Meta) => {
    const videoId = url.pathname.split('/').slice(-1)[0] as string;
    const normalizedUrl = new URL(`http://dood.to/e/${videoId}`);

    const html = await this.fetcher.text(ctx, new URL(normalizedUrl));

    const passMd5Match = html.match(/\/pass_md5\/[\w-]+\/([\w-]+)/);
    if (!passMd5Match) {
      throw new NotFoundError();
    }

    const token = passMd5Match[1] as string;

    const baseUrl = await this.fetcher.text(ctx, new URL(`http://dood.to${passMd5Match[0]}`));

    const $ = cheerio.load(html);
    const title = $('title').text().trim().replace(/ - DoodStream$/, '').trim();

    return {
      url: new URL(`${baseUrl}${randomstring.generate(10)}?token=${token}&expiry=${Date.now()}`),
      label: this.label,
      sourceId: `${this.id}_${meta.countryCode.toLowerCase()}`,
      ttl: this.ttl,
      meta: {
        title,
        ...meta,
      },
      requestHeaders: {
        Referer: 'http://dood.to/',
      },
    };
  };
}
