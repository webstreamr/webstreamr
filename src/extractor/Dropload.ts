import bytes from 'bytes';
import * as cheerio from 'cheerio';
import { Extractor } from './types';
import { extractUrlFromPacked, Fetcher } from '../utils';
import { Context, Meta } from '../types';
import { NotFoundError } from '../error';

export class Dropload implements Extractor {
  readonly id = 'dropload';

  readonly label = 'Dropload';

  readonly ttl = 900000; // 15m

  private readonly fetcher: Fetcher;

  constructor(fetcher: Fetcher) {
    this.fetcher = fetcher;
  }

  readonly supports = (_ctx: Context, url: URL): boolean => null !== url.host.match(/dropload/);

  readonly extract = async (ctx: Context, url: URL, meta: Meta) => {
    url.pathname = url.pathname.replace('/e/', '').replace('/embed-', '/');
    const html = await this.fetcher.text(ctx, url);

    if (html.includes('File Not Found')) {
      throw new NotFoundError();
    }

    const heightMatch = html.match(/\d{3,}x(\d{3,}),/) as string[];

    const sizeMatch = html.match(/([\d.]+ ?[GM]B)/) as string[];

    const $ = cheerio.load(html);
    const title = $('.videoplayer h1').text().trim();

    return {
      url: extractUrlFromPacked(html, [/sources:\[{file:"(.*?)"/]),
      label: this.label,
      sourceId: `${this.id}_${meta.countryCode.toLowerCase()}`,
      ttl: this.ttl,
      meta: {
        bytes: bytes.parse(sizeMatch[1] as string) as number,
        height: parseInt(heightMatch[1] as string) as number,
        title,
        ...meta,
      },
    };
  };
}
