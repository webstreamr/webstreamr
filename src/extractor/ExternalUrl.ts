import { Extractor } from './types';
import { Fetcher } from '../utils';
import { Context, Meta } from '../types';

export class ExternalUrl implements Extractor {
  readonly id = 'external';

  readonly label = 'External';

  readonly ttl = 3600000; // 1h

  private readonly fetcher: Fetcher;

  constructor(fetcher: Fetcher) {
    this.fetcher = fetcher;
  }

  readonly supports = (ctx: Context, url: URL): boolean => !('excludeExternalUrls' in ctx.config) && null !== url.host.match(/.*/);

  readonly extract = async (ctx: Context, url: URL, meta: Meta) => {
    // We only want to make sure that the URL is accessible
    await this.fetcher.head(ctx, url);

    return {
      url: url,
      isExternal: true,
      label: `${url.host}`,
      sourceId: `${this.id}_${meta.countryCode.toLowerCase()}`,
      ttl: this.ttl,
      meta,
    };
  };
}
