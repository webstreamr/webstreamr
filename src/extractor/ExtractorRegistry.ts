import TTLCache from '@isaacs/ttlcache';
import winston from 'winston';
import { Extractor } from './types';
import { Context, Meta, UrlResult } from '../types';
import { Fetcher } from '../utils';
import { DoodStream } from './DoodStream';
import { Dropload } from './Dropload';
import { SuperVideo } from './SuperVideo';
import { ExternalUrl } from './ExternalUrl';
import { NotFoundError } from '../error';

export class ExtractorRegistry {
  private readonly logger: winston.Logger;
  private readonly extractors: Extractor[];
  private readonly urlResultCache: TTLCache<string, UrlResult | undefined>;

  constructor(logger: winston.Logger, fetcher: Fetcher) {
    this.logger = logger;
    this.extractors = [
      new DoodStream(fetcher),
      new Dropload(fetcher),
      new SuperVideo(fetcher),
      new ExternalUrl(fetcher), // fallback extractor which must come last
    ];
    this.urlResultCache = new TTLCache({ max: 1024 });
  }

  readonly handle = async (ctx: Context, url: URL, meta: Meta): Promise<UrlResult | undefined> => {
    let urlResult = this.urlResultCache.get(url.href);
    if (this.urlResultCache.has(url.href)) {
      return urlResult ? { ...urlResult, ttl: this.urlResultCache.getRemainingTTL(url.href) } : undefined;
    }

    const extractor = this.extractors.find(extractor => extractor.supports(ctx, url));
    if (!extractor) {
      return undefined;
    }

    this.logger.info(`Extract stream URL using ${extractor.id} extractor from ${url}`, ctx);

    try {
      urlResult = await extractor.extract(ctx, url, meta);
    } catch (error) {
      if (error instanceof NotFoundError) {
        this.urlResultCache.set(url.href, urlResult, { ttl: extractor.ttl });

        return undefined;
      }

      return {
        url,
        isExternal: true,
        error,
        label: url.host,
        sourceId: `${extractor.id}`,
        ttl: extractor.ttl,
        meta,
      };
    }

    this.urlResultCache.set(url.href, urlResult, { ttl: extractor.ttl });

    return urlResult;
  };
}
