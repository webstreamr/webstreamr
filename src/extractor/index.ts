import { Fetcher } from '../utils';
import { DoodStream } from './DoodStream';
import { Dropload } from './Dropload';
import { ExternalUrl } from './ExternalUrl';
import { Extractor } from './Extractor';
import { Fastream } from './Fastream';
import { FileLions } from './FileLions';
import { HubCloud } from './HubCloud';
import { KinoGer } from './KinoGer';
import { Mixdrop } from './Mixdrop';
import { SaveFiles } from './SaveFiles';
import { Soaper } from './Soaper';
import { StreamEmbed } from './StreamEmbed';
import { Streamtape } from './Streamtape';
import { SuperVideo } from './SuperVideo';
import { Uqload } from './Uqload';
import { VidSrc } from './VidSrc';
import { VixSrc } from './VixSrc';
import { XPrime } from './XPrime';
import { YouTube } from './YouTube';

export * from './Extractor';
export * from './ExtractorRegistry';

export const createExtractors = (fetcher: Fetcher): Extractor[] => [
  new DoodStream(fetcher),
  new Dropload(fetcher),
  new Fastream(fetcher),
  new FileLions(fetcher),
  new HubCloud(fetcher),
  new KinoGer(fetcher),
  new Mixdrop(fetcher),
  new SaveFiles(fetcher),
  new Soaper(fetcher),
  new StreamEmbed(fetcher),
  new Streamtape(fetcher),
  new SuperVideo(fetcher),
  new Uqload(fetcher),
  new VidSrc(fetcher, ['in', 'net', 'xyz', 'io', 'vc']), // https://vidsrc.domains/
  new VixSrc(fetcher),
  new XPrime(fetcher),
  new YouTube(fetcher),
  new ExternalUrl(fetcher), // fallback extractor which must come last
];
