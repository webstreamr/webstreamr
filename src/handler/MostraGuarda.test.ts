import winston from 'winston';
import { MostraGuarda } from './MostraGuarda';
import { Fetcher, ImdbId } from '../utils';
import { ExtractorRegistry } from '../extractor';
import { Context } from '../types';
jest.mock('../utils/Fetcher');

const logger = winston.createLogger({ transports: [new winston.transports.Console({ level: 'nope' })] });
// @ts-expect-error No constructor args needed
const fetcher = new Fetcher();
const ctx: Context = { id: 'id', ip: '127.0.0.1', config: { it: 'on' } };

describe('MostraGuarda', () => {
  let handler: MostraGuarda;

  beforeEach(() => {
    handler = new MostraGuarda(fetcher, new ExtractorRegistry(logger, fetcher));
  });

  test('handles non-existent movies gracefully', async () => {
    const streams = await handler.handle(ctx, 'movie', new ImdbId('tt12345678', undefined, undefined));
    expect(streams).toHaveLength(0);
  });

  test('handle imdb the devil\'s bath', async () => {
    const streams = await handler.handle(ctx, 'movie', new ImdbId('tt29141112', undefined, undefined));
    expect(streams).toMatchSnapshot();
  });
});
