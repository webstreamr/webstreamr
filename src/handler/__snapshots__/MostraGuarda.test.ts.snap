// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MostraGuarda handle imdb the devil's bath 1`] = `
[
  {
    "label": "SuperVideo",
    "meta": {
      "bytes": 1181116006,
      "countryCode": "it",
      "height": 720,
      "title": "the devils bath 2024 sub ita",
    },
    "sourceId": "supervideo_it",
    "ttl": 900000,
    "url": "https://hfs279.serversicuro.cc/hls/,dnzpejj427g4a3gyvbth53rxqewqgqqzp3wr6zwutysvud7zroar3cb75d3q,.urlset/master.m3u8",
  },
  {
    "label": "Dropload",
    "meta": {
      "bytes": 1181116006,
      "countryCode": "it",
      "height": 720,
      "title": "the-devils-bath-2024-sd-sub-ita",
    },
    "sourceId": "dropload_it",
    "ttl": 900000,
    "url": "https://srv34.dropload.io/hls2/01/00200/xsr90y2ltdyx_h/master.m3u8?t=avFOZ-10hDxo6__iaoSUrq7o4yY3ldLw6XDlXxpXF8E&s=1747059001&e=14400&f=1004116&srv=srv24&i=0.0&sp=0&ii=130.61.236.204&p1=srv24&p2=srv24",
  },
  {
    "isExternal": true,
    "label": "mixdrop.ag",
    "meta": {
      "countryCode": "it",
    },
    "sourceId": "external_it",
    "ttl": 3600000,
    "url": "https://mixdrop.ag/e/vk196d6xfzwwo1",
  },
  {
    "label": "DoodStream",
    "meta": {
      "countryCode": "it",
      "title": "the-devils-bath-2024-sd-sub-ita",
    },
    "requestHeaders": {
      "Referer": "http://dood.to/",
    },
    "sourceId": "doodstream_it",
    "ttl": 900000,
    "url": "https://kk892as.cloudatacdn.com/u5kj7j7s27d3sdgge5woezkbi4pu672wktq3aqujw47rbpl3xog2gtqnw2xa/awn9rgotuu~mocked-random-string?token=aw2v2d8uhbrj2ky54d573ujk&expiry=639837296000",
  },
]
`;
